pandas/_libs/index.pyx

```diff
@@ -599,70 +600,91 @@ cpdef convert_scalar(ndarray arr, object value):
-    but use the IndexEngine for computation
-
-    This provides good performance with samller MI's
+    Base class for MultiIndexUIntEngine and MultiIndexPyIntEngine, which
+    represent each label in a MultiIndex as an integer, by juxtaposing the bits
+    encoding each level, with appropriate offsets.
+    Consequently, labels are located by first locating each component on the
+    respective level, then locating for (the integer representation of) codes.
     """
-    def get_indexer(self, values):
-        # convert a MI to an ndarray
-        if hasattr(values, 'values'):
-            values = values.values
-        return super(MultiIndexObjectEngine, self).get_indexer(values)
-
-    cpdef get_loc(self, object val):
-
-        # convert a MI to an ndarray
-        if hasattr(val, 'values'):
-            val = val.values
-        return super(MultiIndexObjectEngine, self).get_loc(val)
+    def __init__(self, object levels, object labels,
+                 ndarray[uint64_t, ndim=1] offsets):
+        self.levels = levels
+        self.offsets = offsets
+
+        # Transform labels in a single array, and add 1 so that we are working
+        # with positive integers (-1 for NaN becomes 0):
+        codes = (np.array(labels, dtype='int64').T + 1).astype('uint64')
+        # Map each codes combination in the index to an integer unambiguously
+        # (no collisions possible), based on the "offsets", which describe the
+        # number of bits to switch labels for each level:
+        lab_ints = self.codes_to_ints(codes)
+
+        # Initialize underlying index
+        self._base.__init__(self, lambda: lab_ints, len(lab_ints))
+
+    def get_indexer(self, object target, object method=None,
+                    object limit=None):
+        level_codes = [self.levels[lev].get_indexer(codes, method=method) + 1
+                       for lev, codes in enumerate(zip(*target))]
+
+        keys_int = self.codes_to_ints(np.array(level_codes, dtype='uint64').T)
+
+        # All methods (exact, backfill, pad) directly map to the respective
+        # methods of the underlying (integers) index...
+        if method is not None:
+            # but underlying backfill and pad methods require index and keys
+            # to be sorted. The index already is (checked in
+            # Index._get_fill_indexer), so sort keys (integer representations):
+            order = np.argsort(keys_int)
+            keys_int = keys_int[order]
+            sup_meth = getattr(self._base, 'get_{}_indexer'.format(method))
+            indexer = sup_meth(self, keys_int, limit=limit)
+            indexer = indexer[order]
+        else:
+            indexer = self._base.get_indexer(self, keys_int)
 
+        return indexer
 
-cdef class MultiIndexHashEngine(ObjectEngine):
-    """
-    Use a hashing based MultiIndex impl
-    but use the IndexEngine for computation
+    def get_loc(self, object key):
+        if is_definitely_invalid_key(key):
```

Comment:
I think we can get rid if this line? (not sure how much perf improvement it would give though, I suppose it is only minor as it is also used in every underling level engine) 
as later on you also check `isinstance(key, tuple)`, which is a stricter check than the above.

That said, that does me wonder if requiring the key to be a tuple is then not a regression?

Reply:
you can prob just catch TypeError, ValueError in the try/except

Reply:
We can talk about which errors to raise where, but in this PR I need to keep behavior unchanged (except bugs I fix) if I want to be sure everything works - i.e. ``TypeError`` vs. ``KeyError``. The code which handles these errors e.g. in ``Index.get_value`` is a huge huge huge mess (once we do remove it, things will be already better). So, for now, I just copy other engines as much as possible.

Reply:
Ah sorry, missed that it raised a different error. Agreed that you shouldn't start changing the error types here. 
But, it seems that, because you are now relying under the hood on actual index engines, the checks also happen there, and if you out leave the check here, you still get a TypeError (only with an error message about the invalid part of the key instead of the full key).

From my timing, removing this one saves around another 1 µs (which is of course not a lot, but in the goal of getting 20 µs towards 5 µs, it is something)